### PR TITLE
is_palindrome function

### PR DESCRIPTION
**Pull request by sashour82**
coding is_palindrome function and its unit test

---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [ ] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/is_palindrome.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+A module for checking if a given string is a palindrome (reads the same backward as forward)
+
+Module contents:
+    - palindrome: input string is palindrome if it reads the same backward as forward
+
+Created on XX XX XX
```
Comment (FalaqMajeed) [inline solutions/is_palindrome.py:9]: Hey, I noticed the creation date is missing. Could you add the date when this file was created?

File: solutions/tests/test_is_palindrome.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    These tests check various cases to ensure that the is_palindrome function correctly
+    identifies palindromes, considering case insensitivity, non-alphanumeric characters,
+    and other edge cases.
+    """
+
+    def test_palindrome(self):
+        """
+        Test cases where the string is a palindrome.
+        """
+        self.assertTrue(is_palindrome("A man, a plan, a canal, Panama"))
```
Comment (FalaqMajeed) [inline solutions/tests/test_is_palindrome.py:27]: I suggest making one assertion at a time in the tests, each with a clear and helpful name. This will make the tests easier to understand and maintain. Also, it might be helpful to organize the tests into standard cases and edge cases to make them more clear and comprehensive.

File: solutions/tests/test_is_palindrome.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.assertTrue(is_palindrome(""))
+
+    def test_single_character(self):
+        """
+        Test case where the input is a single character, which is always a palindrome.
+        """
+        self.assertTrue(is_palindrome("a"))
+        self.assertTrue(is_palindrome("1"))
+
+    def test_only_non_alphanumeric(self):
```
Comment (FalaqMajeed) [inline solutions/tests/test_is_palindrome.py:51]: Hey, I recommend changing this test to a more defensive approach. You can write something like:
with self.assertRaises(AssertionError):
    is_palindrome(["!@#$$@!"])
This will help ensure that the function properly handles invalid inputs. Additionally, you could add another defensive test for non-string inputs to make sure the function raises an assertion error for those as well. Just remember to implement these assertions in the function file to make the validation consistent.

File: solutions/is_palindrome.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    Examples:
+        >>> is_palindrome("A man, a plan, a canal, Panama")
+        True
+        >>> is_palindrome("hello")
+        False
+        >>> is_palindrome("Madam")
+        True
+    """
+    # Remove non-alphanumeric characters and convert to lowercase
```
Comment (FalaqMajeed) [inline solutions/is_palindrome.py:36]: Hey, I suggest raising an AssertionError for non-string inputs or special characters in the function. This will ensure the function handles defensive cases properly. You can then test these scenarios in the test file, which will help create a clear and strong defensive test. This approach will make the code more robust and prevent unexpected behavior. Example (assert isinstance(s, str), 'Input must be a string')

File: solutions/is_palindrome.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+Module contents:
+    - palindrome: input string is palindrome if it reads the same backward as forward
+
+Created on XX XX XX
+@author: Saad M. Ashour
+"""
+
+
+def is_palindrome(s: str) -> bool:
```
Comment (FalaqMajeed) [inline solutions/is_palindrome.py:14]: I recommend changing the parameter name from 's' to something more descriptive, like 'input_string' or 'string'. This will improve clarity and make the function easier to understand at a glance. A more specific name helps anyone reading the code quickly grasp its purpose.

**Review by FalaqMajeed — CHANGES_REQUESTED** [review]
Your function idea is creative and unique! I’ve suggested a few changes to improve clarity. Please address these changes to ensure the code is as clear as possible. Feel free to reach out if anything is unclear or if you need further clarification. Thanks!

File: solutions/is_palindrome.py (hunk 1 above)
Comment (sashour82) [reply solutions/is_palindrome.py:9]: done

File: solutions/tests/test_is_palindrome.py (hunk 2 above)
Comment (sashour82) [reply solutions/tests/test_is_palindrome.py:27]: done

File: solutions/is_palindrome.py (hunk 4 above)
Comment (sashour82) [reply solutions/is_palindrome.py:36]: done

File: solutions/is_palindrome.py (hunk 5 above)
Comment (sashour82) [reply solutions/is_palindrome.py:14]: done

**Review by FalaqMajeed — APPROVED** [review]
Thank you for considering my suggestion and updating the code! The changes look fantastic and align perfectly with the checklist requirements.
